### PR TITLE
138945107 extract api service v5

### PR DESCRIPTION
**Pull request by tmikeschu**
Worked with @wlffann @kheppenstall

This PR extracts all API requests to a new model, PivotalTracker::ApiService

Some query methods are refactored so the ApiService methods can be more generalized.

Rebased and ready to rock. 🎸

### CODE REVIEW

File: lib/pivotal-tracker/account_membership.rb [hunk 1]
```diff
@@ -5,22 +5,12 @@ class PivotalTracker::AccountMembership < PivotalTracker::Resource
...
   def self.all(account_id)
-    response = PivotalTracker.get("/accounts/#{account_id}/memberships")		
-    if response.code == 200		
-      response.parsed_response.map do |membership|		
-        PivotalTracker::AccountMembership.new(membership)		
-      end		
-    else		
-      raise(PivotalTracker::PermissionDenied, "Only Admins and Owners can see account memberships")		
-    end		
+    memberships = PivotalTracker::ApiService.all_nested('accounts', account_id, 'memberships')
```
Comment (jasonnoble) [inline lib/pivotal-tracker/account_membership.rb]: raw_memberships?

Comment (jasonnoble) [reply lib/pivotal-tracker/account_membership.rb]: Or `memberships_json`

File: lib/pivotal-tracker/api_service.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+class PivotalTracker::ApiService
+  def self.find(collection, id)
```
Comment (jasonnoble) [inline lib/pivotal-tracker/api_service.rb]: Should this class be all the API related stuff currently at the top of pivotal_tracker.rb? i.e. HTTParty, base_uri, PermissionDenied error

File: lib/pivotal-tracker/api_service.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    response = PivotalTracker.post("/accounts/#{account_id}/memberships", query: membership_attributes)
+    if success?(response.code)
+      response.parsed_response
+    else
+      raise(PivotalTracker::PermissionDenied, "Only Admins and Owners can add account memberships")
+    end
+  end
+
+  def self.all(collection)
+    PivotalTracker.get("/#{collection}").parsed_response
```
Comment (jasonnoble) [inline lib/pivotal-tracker/api_service.rb]: Add `if_success?` code here...

File: lib/pivotal-tracker/api_service.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  end
+
+  def self.all_nested(owner_collection, owner_id, target_collection)
+    response = PivotalTracker.get("/#{owner_collection}/#{owner_id}/#{target_collection}")
+    if success?(response.code)
+      response.parsed_response
+    else
+      puts response.inspect
```
Comment (jasonnoble) [inline lib/pivotal-tracker/api_service.rb]: Hmmm.

File: lib/pivotal-tracker/api_service.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    end
+  end
+
+  def self.get_token(username, password)
+    PivotalTracker.basic_auth username, password
```
Comment (jasonnoble) [inline lib/pivotal-tracker/api_service.rb]: If we do https://github.com/jasonnoble/pivotal-tracker/pull/33/files#r99225728, this should just be `basic_auth username, password`.

File: lib/pivotal-tracker/api_service.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    PivotalTracker.get('/me')['api_token']
```
Comment (jasonnoble) [inline lib/pivotal-tracker/api_service.rb]: Remember, the `PivotalTracker` class shouldn't be doing `get` any more.  Should this be `me`?  i.e. `PivotalTracker::Api.me`?

File: lib/pivotal-tracker/project.rb [hunk 7]
```diff
@@ -9,25 +9,16 @@ class PivotalTracker::Project < PivotalTracker::Resource
                 :public, :start_date, :start_time, :time_zone, :updated_at,
                 :velocity_averaged_over, :version, :week_start_day
   def self.all
-    PivotalTracker.get('/projects').parsed_response.map{|project| PivotalTracker::Project.new(project) }
+    projects = PivotalTracker::ApiService.all('projects')
```
Comment (jasonnoble) [inline lib/pivotal-tracker/project.rb]: `projects_json`?

**Review by jasonnoble — CHANGES_REQUESTED** [review]

File: lib/pivotal-tracker/project.rb (hunk 7 above)
Comment (tmikeschu) [reply lib/pivotal-tracker/project.rb]: 👍

File: lib/pivotal-tracker/account_membership.rb (hunk 1 above)
Comment (tmikeschu) [reply lib/pivotal-tracker/account_membership.rb]: 👍

File: lib/pivotal-tracker/api_service.rb (hunk 2 above)
Comment (tmikeschu) [reply lib/pivotal-tracker/api_service.rb]: 👍

**Pull request merged**